libproxy/modules/pacrunner_duktape.cpp

```diff
@@ -0,0 +1,137 @@
+ * libproxy - A library for proxy configuration
+ * Copyright (C) 2006 Nathaniel McCallum <nathaniel@natemccallum.com>
+ * Copyright (C) 2021 Zhaofeng Li <hello@zhaofeng.li>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ ******************************************************************************/
+
+#include "../extension_pacrunner.hpp"
+#include <unistd.h> // gethostname
+using namespace libproxy;
+
+#include <duktape.h>
+#include "pacutils.h"
+
+static duk_ret_t dnsResolve(duk_context *ctx) {
+	if (duk_get_top(ctx) != 1) {
+		// Invalid number of arguments
+		return 0;
+	}
+
+	// We do not need to free the string - It's managed by Duktape.
+	const char *hostname = duk_get_string(ctx, 0);
+	if (!hostname) {
+		return 0;
+	}
+
+	// Look it up
+	struct addrinfo *info;
+	if (getaddrinfo(hostname, NULL, NULL, &info)) {
+		return 0;
+	}
+
+	// Try for IPv4
+	char *tmp = new char[INET6_ADDRSTRLEN+1];
+	if (getnameinfo(info->ai_addr, info->ai_addrlen,
+					tmp, INET6_ADDRSTRLEN+1,
+					NULL, 0,
+					NI_NUMERICHOST)) {
+		freeaddrinfo(info);
+		delete[] tmp;
+		return 0;
+	}
+
+	// Create the return value
+	duk_push_string(ctx, tmp);
+	delete[] tmp;
+	return 1;
+}
+
+static duk_ret_t myIpAddress(duk_context *ctx) {
+	char hostname[1024];
+	if (!gethostname(hostname, 1023)) {
+		duk_push_string(ctx, hostname);
+		return dnsResolve(ctx);
+	}
+
+	return duk_error(ctx, DUK_ERR_ERROR, "Unable to find hostname!");
+}
+
+class duktape_pacrunner : public pacrunner {
+public:
+	duktape_pacrunner(string pac, const url& pacurl) : pacrunner(pac, pacurl) {
+		duk_context *ctx = duk_create_heap_default();
+		if (!ctx) goto error;
+		duk_push_c_function(ctx, dnsResolve, 1);
+		duk_put_global_string(ctx, "dnsResolve");
+
+		duk_push_c_function(ctx, myIpAddress, 1);
+		duk_put_global_string(ctx, "myIpAddress");
+
+		// Add other routines
+		duk_push_string(ctx, JAVASCRIPT_ROUTINES);
+		if (duk_peval_noresult(ctx)) {
+			goto error;
+		}
+
+		// Add the PAC into the context
+		duk_push_string(ctx, pac.c_str());
+		if (duk_peval_noresult(ctx)) {
+			goto error;
+		}
+
+		this->ctx = ctx;
+		return;
+	error:
+		duk_destroy_heap(ctx);
+		throw bad_alloc();
+	}
+
+	~duktape_pacrunner() {
+		if (this->ctx) duk_destroy_heap(ctx);
```

Comment:
Can never be `nullptr`

```suggestion
		duk_destroy_heap(ctx);
```